File: backend/app/repositories/telemetry.py
"""Telemetry storage and retrieval."""

from __future__ import annotations

import logging
from datetime import datetime, timedelta
from typing import Any

from google.cloud.firestore_v1 import FieldFilter

from app.firebase.client import Collections
from app.models.domain import (
    Bucket,
    ReadingQuality,
    ReadingSource,
    SensorReading,
    SensorType,
    TimeSeries,
    TimeSeriesPoint,
)
from app.repositories.base import db, snapshot_data, to_datetime, to_float

logger = logging.getLogger(__name__)

LatestReadings = dict[SensorType, SensorReading]


def _map_reading(data: dict[str, Any], doc_id: str) -> SensorReading | None:
    """A reading without a value, a timestamp or a type is not a reading."""
    value = to_float(data.get("value"))
    timestamp = to_datetime(data.get("timestamp"))
    sensor_type = data.get("sensorType")

    if value is None or timestamp is None or sensor_type not in {s.value for s in SensorType}:
        logger.debug("Discarding malformed reading %s", doc_id)
        return None

    return SensorReading(
        id=doc_id,
        farm_id=str(data.get("farmId") or ""),
        zone_id=str(data["zoneId"]) if data.get("zoneId") else None,
        sensor_id=str(data.get("sensorId") or ""),
        sensor_type=SensorType(sensor_type),
        value=value,
        unit=str(data.get("unit") or ""),
        timestamp=timestamp,
        quality=ReadingQuality(data.get("quality") or ReadingQuality.GOOD.value),
        source=ReadingSource(data.get("source") or ReadingSource.ESP32.value),
    )


def _latest_doc_id(farm_id: str, zone_id: str | None, sensor_type: SensorType) -> str:
    return f"{farm_id}__{zone_id or 'farm'}__{sensor_type.value}"


def write_reading(reading: SensorReading) -> SensorReading:
    """Append to history and upsert the latest-value document in one batch."""
    payload = {
        "farmId": reading.farm_id,
        "zoneId": reading.zone_id,
        "sensorId": reading.sensor_id,
        "sensorType": reading.sensor_type.value
        if isinstance(reading.sensor_type, SensorType)
        else reading.sensor_type,
        "value": reading.value,
        "unit": reading.unit,
        "timestamp": reading.timestamp,
        "quality": reading.quality,
        "source": reading.source,
    }

    batch = db().batch()
    history_ref = db().collection(Collections.SENSOR_READINGS).document()
    batch.set(history_ref, payload)

    latest_ref = db().collection(Collections.LATEST_READINGS).document(
        _latest_doc_id(reading.farm_id, reading.zone_id, SensorType(payload["sensorType"]))
    )
    batch.set(latest_ref, payload, merge=True)
    batch.commit()

    return reading.model_copy(update={"id": history_ref.id})


def get_latest_readings(farm_id: str, zone_id: str | None = None) -> LatestReadings:
    query = (
        db()
        .collection(Collections.LATEST_READINGS)
        .where(filter=FieldFilter("farmId", "==", farm_id))
    )
    if zone_id:
        query = query.where(filter=FieldFilter("zoneId", "==", zone_id))

    latest: LatestReadings = {}
    for doc in query.stream():
        reading = _map_reading(snapshot_data(doc), doc.id)
        if reading is None:
            continue
        existing = latest.get(reading.sensor_type)
        if existing is None or reading.timestamp > existing.timestamp:
            latest[reading.sensor_type] = reading
    return latest


def get_latest_readings_by_zone(farm_id: str) -> dict[str, LatestReadings]:
    """Latest readings grouped by zone, for farm-wide evaluation in one query."""
    query = (
        db()
        .collection(Collections.LATEST_READINGS)
        .where(filter=FieldFilter("farmId", "==", farm_id))
    )
    grouped: dict[str, LatestReadings] = {}
    for doc in query.stream():
        reading = _map_reading(snapshot_data(doc), doc.id)
        if reading is None or reading.zone_id is None:
            continue
        bucket = grouped.setdefault(reading.zone_id, {})
        existing = bucket.get(reading.sensor_type)
        if existing is None or reading.timestamp > existing.timestamp:
            bucket[reading.sensor_type] = reading
    return grouped


def get_history(
    farm_id: str,
    sensor_type: SensorType,
    start: datetime,
    end: datetime,
    zone_id: str | None = None,
    bucket: Bucket = Bucket.RAW,
    limit: int = 5000,
) -> TimeSeries:
    query = (
        db()
        .collection(Collections.SENSOR_READINGS)
        .where(filter=FieldFilter("farmId", "==", farm_id))
        .where(filter=FieldFilter("sensorType", "==", sensor_type.value))
        .where(filter=FieldFilter("timestamp", ">=", start))
        .where(filter=FieldFilter("timestamp", "<=", end))
    )
    if zone_id:
        query = query.where(filter=FieldFilter("zoneId", "==", zone_id))

    readings: list[SensorReading] = []
    for doc in query.order_by("timestamp").limit(limit).stream():
        mapped = _map_reading(snapshot_data(doc), doc.id)
        if mapped is not None:
            readings.append(mapped)

    points = [TimeSeriesPoint(timestamp=r.timestamp, value=r.value) for r in readings]
    if bucket is not Bucket.RAW:
        points = bucket_points(points, bucket)

    return TimeSeries(
        sensor_type=sensor_type,
        zone_id=zone_id,
        unit=readings[0].unit if readings else "",
        points=points,
    )


def bucket_points(points: list[TimeSeriesPoint], bucket: Bucket) -> list[TimeSeriesPoint]:
    """Average points into hour, day or month buckets. Empty stays empty."""
    if not points:
        return []

    groups: dict[datetime, list[float]] = {}
    for point in points:
        ts = point.timestamp
        if bucket is Bucket.HOUR:
            key = ts.replace(minute=0, second=0, microsecond=0)
        elif bucket is Bucket.DAY:
            key = ts.replace(hour=0, minute=0, second=0, microsecond=0)
        else:
            key = ts.replace(day=1, hour=0, minute=0, second=0, microsecond=0)
        groups.setdefault(key, []).append(point.value)

    return [
        TimeSeriesPoint(timestamp=key, value=sum(values) / len(values))
        for key, values in sorted(groups.items())
    ]


def sum_over_window(
    farm_id: str,
    sensor_type: SensorType,
    start: datetime,
    end: datetime,
    zone_id: str | None = None,
) -> tuple[float | None, str]:
    """Mean of stored readings in a window, or None when the window is empty."""
    series = get_history(farm_id, sensor_type, start, end, zone_id=zone_id)
    if not series.points:
        return None, series.unit
    return sum(p.value for p in series.points) / len(series.points), series.unit


def window_for(period: str, reference: datetime) -> tuple[datetime, datetime, datetime, datetime]:
    """Current and previous window bounds for a comparison period."""
    day_start = reference.replace(hour=0, minute=0, second=0, microsecond=0)

    if period in {"TODAY", "DAY"}:
        return day_start, reference, day_start - timedelta(days=1), day_start
    if period == "WEEK":
        week_start = day_start - timedelta(days=day_start.weekday())
        return week_start, reference, week_start - timedelta(days=7), week_start
    if period == "MONTH":
        month_start = day_start.replace(day=1)
        previous_end = month_start
        previous_start = (month_start - timedelta(days=1)).replace(day=1)
        return month_start, reference, previous_start, previous_end
    year_start = day_start.replace(month=1, day=1)
    return year_start, reference, year_start.replace(year=year_start.year - 1), year_start

File: src/services/firebase/telemetry.ts
import {
  collection,
  doc,
  getDoc,
  getDocs,
  limit as fbLimit,
  onSnapshot,
  orderBy,
  query,
  where,
} from 'firebase/firestore';
import { COLLECTIONS, getDb } from '@/firebase/config';
import { toISO, toNumber } from '@/firebase/converters';
import type { TelemetryRepository, Unsubscribe } from '@/services/repositories';
import type {
  ComparisonResult,
  LatestReadings,
  SensorReading,
  SensorType,
  TimeSeries,
  TimeSeriesPoint,
} from '@/types';

function mapReading(data: Record<string, unknown>, id: string): SensorReading | null {
  const value = toNumber(data.value);
  const timestamp = toISO(data.timestamp);
  const sensorType = data.sensorType as SensorType | undefined;
  // A reading without a value, a time or a type is not a reading. Drop it
  // rather than displaying a placeholder number.
  if (value === undefined || !timestamp || !sensorType) return null;
  return {
    id,
    farmId: String(data.farmId ?? ''),
    zoneId: data.zoneId ? String(data.zoneId) : undefined,
    sensorId: String(data.sensorId ?? ''),
    sensorType,
    value,
    unit: String(data.unit ?? ''),
    timestamp,
    quality: (data.quality as SensorReading['quality']) ?? 'GOOD',
    source: (data.source as SensorReading['source']) ?? 'ESP32',
  };
}

function collapseToLatest(readings: SensorReading[]): LatestReadings {
  const latest: LatestReadings = {};
  for (const reading of readings) {
    const existing = latest[reading.sensorType];
    if (!existing || new Date(reading.timestamp) > new Date(existing.timestamp)) {
      latest[reading.sensorType] = reading;
    }
  }
  return latest;
}

export class FirebaseTelemetryRepository implements TelemetryRepository {
  private latestQuery(farmId: string, zoneId?: string) {
    const base = collection(getDb(), COLLECTIONS.latestReadings);
    return zoneId
      ? query(base, where('farmId', '==', farmId), where('zoneId', '==', zoneId))
      : query(base, where('farmId', '==', farmId));
  }

  async getLatestReadings(farmId: string, zoneId?: string): Promise<LatestReadings> {
    const snapshot = await getDocs(this.latestQuery(farmId, zoneId));
    const readings = snapshot.docs
      .map((d) => mapReading(d.data(), d.id))
      .filter((r): r is SensorReading => r !== null);
    return collapseToLatest(readings);
  }

  subscribeToLatestReadings(
    farmId: string,
    zoneId: string | undefined,
    onChange: (readings: LatestReadings) => void,
    onError: (e: Error) => void,
  ): Unsubscribe {
    return onSnapshot(
      this.latestQuery(farmId, zoneId),
      (snapshot) => {
        const readings = snapshot.docs
          .map((d) => mapReading(d.data(), d.id))
          .filter((r): r is SensorReading => r !== null);
        onChange(collapseToLatest(readings));
      },
      (error) => onError(error as Error),
    );
  }

  async getHistory(params: {
    farmId: string;
    zoneId?: string;
    sensorType: SensorType;
    from: string;
    to: string;
    bucket?: 'raw' | 'hour' | 'day' | 'month';
  }): Promise<TimeSeries> {
    const constraints = [
      where('farmId', '==', params.farmId),
      where('sensorType', '==', params.sensorType),
      where('timestamp', '>=', new Date(params.from)),
      where('timestamp', '<=', new Date(params.to)),
      orderBy('timestamp', 'asc'),
      fbLimit(5000),
    ];
    if (params.zoneId) constraints.unshift(where('zoneId', '==', params.zoneId));

    const snapshot = await getDocs(query(collection(getDb(), COLLECTIONS.sensorReadings), ...constraints));
    const readings = snapshot.docs
      .map((d) => mapReading(d.data(), d.id))
      .filter((r): r is SensorReading => r !== null);

    const points: TimeSeriesPoint[] = readings.map((r) => ({ timestamp: r.timestamp, value: r.value }));
    return {
      sensorType: params.sensorType,
      zoneId: params.zoneId,
      unit: readings[0]?.unit ?? '',
      points: params.bucket && params.bucket !== 'raw' ? bucketPoints(points, params.bucket) : points,
    };
  }

  /**
   * Comparison is derived only from stored readings. When either period holds no
   * data, sufficientData is false and the UI shows the insufficient-history
   * message instead of a number.
   */
  async getComparison(params: {
    farmId: string;
    zoneId?: string;
    metric: Parameters<TelemetryRepository['getComparison']>[0]['metric'];
    period: ComparisonResult extends never ? never : Parameters<TelemetryRepository['getComparison']>[0]['period'];
  }): Promise<ComparisonResult> {
    // Aggregate documents are written by the backend; the client never computes
    // a comparison from data it invented.
    const id = [params.farmId, params.zoneId ?? 'farm', params.metric, params.period].join('__');
    const snapshot = await getDoc(doc(getDb(), 'comparisons', id));

    if (!snapshot.exists()) {
      return {
        metric: String(params.metric),
        unit: '',
        currentValue: null,
        previousValue: null,
        absoluteDifference: null,
        percentageDifference: null,
        trend: 'UNKNOWN',
        sufficientData: false,
      };
    }

    const data = snapshot.data();
    const current = toNumber(data.currentValue) ?? null;
    const previous = toNumber(data.previousValue) ?? null;
    const sufficient = current !== null && previous !== null;
    const absolute = sufficient ? current - previous : null;
    const percentage = sufficient && previous !== 0 ? ((current - previous) / Math.abs(previous)) * 100 : null;

    return {
      metric: String(params.metric),
      unit: String(data.unit ?? ''),
      currentValue: current,
      previousValue: previous,
      absoluteDifference: absolute,
      percentageDifference: percentage,
      trend: absolute === null ? 'UNKNOWN' : absolute > 0 ? 'UP' : absolute < 0 ? 'DOWN' : 'FLAT',
      sufficientData: sufficient,
    };
  }
}

/** Averages readings into hour, day or month buckets. Empty input stays empty. */
function bucketPoints(points: TimeSeriesPoint[], bucket: 'hour' | 'day' | 'month'): TimeSeriesPoint[] {
  const groups = new Map<string, number[]>();
  for (const point of points) {
    const date = new Date(point.timestamp);
    const key =
      bucket === 'hour'
        ? `${date.toISOString().slice(0, 13)}:00:00.000Z`
        : bucket === 'day'
          ? `${date.toISOString().slice(0, 10)}T00:00:00.000Z`
          : `${date.toISOString().slice(0, 7)}-01T00:00:00.000Z`;
    const list = groups.get(key);
    if (list) list.push(point.value);
    else groups.set(key, [point.value]);
  }
  return [...groups.entries()]
    .map(([timestamp, values]) => ({
      timestamp,
      value: values.reduce((sum, v) => sum + v, 0) / values.length,
    }))
    .sort((a, b) => a.timestamp.localeCompare(b.timestamp));
}
